### PR TITLE
Introduce Tracing Annotations

### PR DESCRIPTION
**Pull request by ihcsim**
This PR introduces two proxy config annotations named `config.linkerd.io/trace-collector` and `config.alpha.linkerd.io/trace-collector-service-account`, to support per-pod tracing. These annotations are translated into two new proxy environment variables.

Note that these annotations work on both the pod template and namespace levels.

The proxy injector logs and events have been updated to include `tracing enabled` events.
```
linkerd-proxy-injector-667958849d-pkvkn proxy-injector time="2019-09-26T04:32:23Z" level=info msg="patch generated for: pod/voting-855d545684-"
linkerd-proxy-injector-667958849d-pkvkn proxy-injector time="2019-09-26T04:32:23Z" level=info msg="received admission review request 93017758-75b4-4aac-8721-c9c37d11b043"
linkerd-proxy-injector-667958849d-pkvkn proxy-injector time="2019-09-26T04:32:23Z" level=info msg="received pod/web-79694f4896-"
linkerd-proxy-injector-667958849d-pkvkn proxy-injector time="2019-09-26T04:32:23Z" level=info msg="tracing enabled for deployment/web: remote service=oc-collector.tracing:55678, service account=tracing.tracing"


Events:
  Type    Reason             Age   From                    Message
  ----    ------             ----  ----                    -------
  Normal  ScalingReplicaSet  83s   deployment-controller   Scaled up replica set emoji-697b575bd9 to 1
  Normal  Injected           83s   linkerd-proxy-injector  Linkerd sidecar proxy injected
  Normal  Tracing            83s   linkerd-proxy-injector  Tracing Enabled
```

Fixes https://github.com/linkerd/linkerd2/issues/3472.

### CODE REVIEW

File: pkg/inject/inject.go [hunk 1]
```diff
@@ -539,6 +546,32 @@ func (conf *ResourceConfig) serviceAccountVolumeMount() *corev1.VolumeMount {
...
+		svcAccount = conf.getOverride(k8s.ProxyTraceCollectorSvcAccount)
+	)
+
+	if svcAddr == "" {
+		return nil
+	}
+
+	splits := strings.Split(svcAddr, ".")
+	if len(splits) <= 1 {
+		log.Errorf("fail to enable tracing due to malformed collector service address %s. Expected service address to use the <svc-name>:<ns> format", svcAddr)
```
Comment (alpeb) [inline pkg/inject/inject.go]: Expected should be `<svc-name>.<ns>`, right?

File: controller/proxy-injector/webhook.go [hunk 2]
```diff
@@ -108,6 +109,9 @@ func Inject(api *k8s.API,
 
 	if parent != nil {
 		recorder.Event(*parent, v1.EventTypeNormal, eventTypeInjected, "Linkerd sidecar proxy injected")
+		if report.TracingEnabled {
```
Comment (grampelberg) [inline controller/proxy-injector/webhook.go]: Should we have a disabled event as well? I'm not 100% sure it is needed.

File: pkg/inject/inject.go (hunk 1 above)
Comment (alpeb) [reply pkg/inject/inject.go]: Also isn't it possible for `config.linkerd.io/trace-collector`'s value to have more than one dot?

File: pkg/charts/values_test.go [hunk 3]
```diff
@@ -85,6 +85,10 @@ func TestNewValues(t *testing.T) {
 					Request: "",
 				},
 			},
+			Trace: &Trace{
+				CollectorSvcAddr:    "",
+				CollectorSvcAccount: "default",
```
Comment (grampelberg) [inline pkg/charts/values_test.go]: It looks like you're expecting the namespace further down, so this won't work?

File: pkg/inject/inject.go [hunk 4]
```diff
@@ -539,6 +546,32 @@ func (conf *ResourceConfig) serviceAccountVolumeMount() *corev1.VolumeMount {
 	return nil
 }
 
+func (conf *ResourceConfig) trace() *charts.Trace {
+	var (
+		svcAddr    = conf.getOverride(k8s.ProxyTraceCollectorSvcAddr)
+		svcAccount = conf.getOverride(k8s.ProxyTraceCollectorSvcAccount)
```
Comment (alpeb) [inline pkg/inject/inject.go]: Shouldn't this be defaulted to `default` if the annotation is missing?

Comment (ihcsim) [reply pkg/inject/inject.go]: The default is read from `values.yaml`, like all other defaults.

File: controller/proxy-injector/webhook.go (hunk 2 above)
Comment (ihcsim) [reply controller/proxy-injector/webhook.go]: Just like inject/skip, I think it will be easier for users to determine if tracing is enabled by looking at events than the proxy injector logs. Happy to remove if you don't think it's needed.

File: pkg/inject/inject.go (hunk 4 above)
Comment (alpeb) [reply pkg/inject/inject.go]: I think you're missing a step. `conf.getOverride()` only checks for annotations. It seems to me you need to add the `Trace` struct into the protobuf def, and then if `conf.getOverride()` doesn't return anything, default to something like `conf.configs.GetProxy().GetTrace().GetCollectorSvcAddr()` and `conf.configs.GetProxy().GetTrace().GetCollectorSvcAccount()`

File: pkg/charts/values_test.go (hunk 3 above)
Comment (ihcsim) [reply pkg/charts/values_test.go]: This test is to validate the defaults read from the `values.yaml` file. The logic that appends the namespace is in the `pkg/inject` package. The test in `pkg/inject/inject_test.go` will require the namespace suffix to be added as seen [here](https://github.com/linkerd/linkerd2/pull/3481/files#diff-61c5e7cb5fc7a736c4ba737d5ad5f4ebR142).

File: controller/proxy-injector/webhook.go (hunk 2 above)
Comment (ihcsim) [reply controller/proxy-injector/webhook.go]: Sorry, completely misread your comment. Yeah, I agree we don't need to add `disable` events, as I like to reserve events for unusual stuff.

File: pkg/inject/inject.go (hunk 4 above)
Comment (ihcsim) [reply pkg/inject/inject.go]: The reason why I didn't update the protobuf is because everything in the pb def (i.e. all the configuration in the `linkerd-config` config map) is meant to apply to *all* proxies, and be persisted across upgrades. Since #3472 is limited to (on demand) enabling tracing per workload, I think we don't need to add these to the pb def.

In this case, if `conf.getOverride()` returns an empty string, implying that no tracing annotations are added, we assume tracing is disabled.

Comment (alpeb) [reply pkg/inject/inject.go]: Ok, in that case what are the `Trace` properties in `values.yaml` for, since we only care for the annotation values?

File: pkg/inject/inject.go (hunk 1 above)
Comment (ihcsim) [reply pkg/inject/inject.go]: It is possible, but this check only logs an error if the provided service address only has the service name, with no namespace or namspace + cluster domain suffix.

File: pkg/inject/inject.go [hunk 5]
```diff
@@ -539,6 +546,32 @@ func (conf *ResourceConfig) serviceAccountVolumeMount() *corev1.VolumeMount {
...
+		return nil
+	}
+
+	splits := strings.Split(svcAddr, ".")
+	if len(splits) <= 1 {
+		log.Errorf("fail to enable tracing due to malformed collector service address %s. Expected service address to use the <svc-name>:<ns> format", svcAddr)
+		return nil
+	}
+
+	if delimiter := strings.Index(splits[1], ":"); delimiter != -1 {
```
Comment (adleong) [inline pkg/inject/inject.go]: What is this doing?  Stripping the port off?  Is there a more clear way to do this, perhaps before splitting the domain into segments?

File: pkg/inject/inject.go (hunk 1 above)
Comment (adleong) [inline pkg/inject/inject.go]: Do we want to support a single segment service name if the service is in the same namespace as the pod?

File: charts/partials/templates/_proxy.tpl [hunk 6]
```diff
@@ -68,6 +68,14 @@ env:
 - name: LINKERD2_PROXY_TAP_SVC_NAME
   value: linkerd-tap.$(_l5d_ns).serviceaccount.identity.$(_l5d_ns).$(_l5d_trustdomain)
 {{ end -}}
+{{ if .Proxy.Trace -}}
+{{ if .Proxy.Trace.CollectorSvcAddr -}}
+- name: LINKERD2_PROXY_TRACE_COLLECTOR_SVC_ADDR
+  value: {{ .Proxy.Trace.CollectorSvcAddr }}
+- name: LINKERD2_PROXY_TRACE_COLLECTOR_SVC_NAME
+  value: {{ .Proxy.Trace.CollectorSvcAccount }}.serviceaccount.identity.$(_l5d_ns).$(_l5d_trustdomain)
```
Comment (adleong) [inline charts/partials/templates/_proxy.tpl]: I found it a bit hard to follow that this variable contains both the service account name and namespace.  I think it would be more clear if there was a template variable for the entire identity string and the value for the whole identity string was composed in the injector code *OR* if there were two separate template variables for service account name and service account namespace.

File: pkg/inject/inject.go (hunk 4 above)
Comment (ihcsim) [reply pkg/inject/inject.go]: Ha.. you are right. It was a leftover from pre- a4b17184add0b3f6b9579a459e32e9ad3aea1bdc.

Comment (ihcsim) [reply pkg/inject/inject.go]: Ha.. you are right. It was a leftover from pre- a4b17184add0b3f6b9579a459e32e9ad3aea1bdc.

File: pkg/inject/report.go [hunk 7]
```diff
@@ -77,6 +78,7 @@ func newReport(conf *ResourceConfig) *Report {
 		report.HostNetwork = conf.pod.spec.HostNetwork
 		report.Sidecar = healthcheck.HasExistingSidecars(conf.pod.spec)
 		report.UDP = checkUDPPorts(conf.pod.spec)
+		report.TracingEnabled = conf.pod.meta.Annotations[k8s.ProxyTraceCollectorSvcAddr] != "" || conf.nsAnnotations[k8s.ProxyTraceCollectorSvcAddr] != ""
```
Comment (adleong) [inline pkg/inject/report.go]: Feel free to ignore for now, but just a forward looking question:

Are reports always based just on the annotations?  If we, in the future, support setting default tracing values in the linkerd-config, is it possible to report that tracing is enabled even when it's not on the annotations?

File: pkg/inject/inject.go (hunk 1 above)
Comment (alpeb) [reply pkg/inject/inject.go]: If the there's more than one dot in that address then the splitting over the next lines might not work as expected. I'll let you discuss with @adleong over his suggestions ;-)

File: pkg/inject/report.go (hunk 7 above)
Comment (ihcsim) [reply pkg/inject/report.go]: Yes, if down the road, we have other ways to enable tracing, we will need to account for that here.

File: pkg/inject/inject.go (hunk 4 above)
Comment (ihcsim) [reply pkg/inject/inject.go]: Per slack chat, we will keep these as they are needed for https://github.com/linkerd/linkerd2/issues/3468.

File: pkg/inject/inject.go (hunk 1 above)
Comment (ihcsim) [reply pkg/inject/inject.go]: Good point. I forgot that the `.ns` suffix is optional if the `Service` is in the same namespace.

Comment (ihcsim) [reply pkg/inject/inject.go]: Good catch on the subsequent call to `strings.Split()`.

Comment (ihcsim) [reply pkg/inject/inject.go]: Since the namespace that will be appended to the service account below is derived from the service address, a single segment service name isn't going to work. We will need to add this to the docs.

Comment (adleong) [reply pkg/inject/inject.go]: if the service namespace isn't provided, isn't it assumed to be the same as the pod?  i.e. if a pod in the `foo` namespace specifies the `bar` service, we would assume that means the `bar` service in the `foo` namespace and also use `foo` as the namespace for the service account?

File: charts/partials/templates/_proxy.tpl (hunk 6 above)
Comment (ihcsim) [reply charts/partials/templates/_proxy.tpl]: For some reasons, I thought that `<serviceaccount>.<ns>` is the fully qualified name of a service account, but turns out it's something like `<group>:<ns>:<serviceaccount>`. I'm gonna keep the `.serviceaccount.identity.$(_l5d_ns).$(_l5d_trustdomain)` as-is to remain consistent with the rest of the `*_SVC_NAME` in this template, and might just change the `CollectorSvcAccount` variable name to something more reflective. Any suggestions on variable name?

File: pkg/inject/inject.go [hunk 8]
```diff
@@ -497,6 +501,11 @@ func (conf *ResourceConfig) injectPodSpec(values *patch) {
 	}
 
 	values.AddRootVolumes = len(conf.pod.spec.Volumes) == 0
+
+	if trace := conf.trace(); trace != nil {
+		log.Infof("tracing enabled for %s/%s: remote service=%s, service account=%s", conf.workload.ownerRef.Kind, conf.workload.ownerRef.Name, trace.CollectorSvcAddr, trace.CollectorSvcAccount)
```
Comment (alpeb) [inline pkg/inject/inject.go]: When running `linkerd inject --manual` against some uninjected payload with the trace annotation, this is throwing a segfault because `conf.workload.ownerRef` is nil.

File: pkg/charts/values_test.go (hunk 3 above)
Comment (adleong) [reply pkg/charts/values_test.go]: I'm trying to understand this.  Does the injector use values.yaml as the defaults for the Values struct?  If so, I think this is a problem.  For any pods that don't have the annotation, these default values would be used and plugged into the proxy partial, leading to a `LINKERD2_PROXY_TRACE_COLLECTOR_SVC_NAME` of `default.serviceaccount.identity.$(_l5d_ns).$(_l5d_trustdomain)` which is missing the namespace.

Comment (adleong) [reply pkg/charts/values_test.go]: Wouldn't it be more appropriate for the Trace struct to be absent from Values.yaml?  So that the default value of `Trace` is nil?

File: pkg/inject/inject.go [hunk 9]
```diff
@@ -539,6 +548,36 @@ func (conf *ResourceConfig) serviceAccountVolumeMount() *corev1.VolumeMount {
 	return nil
 }
 
+func (conf *ResourceConfig) trace() *charts.Trace {
+	var (
+		svcAddr    = conf.getOverride(k8s.ProxyTraceCollectorSvcAddr)
```
Comment (adleong) [inline pkg/inject/inject.go]: just curious why these are var definitions instead of `:=` assignments

File: pkg/inject/inject.go (hunk 8 above)
Comment (ihcsim) [reply pkg/inject/inject.go]: Good catch! Thanks. We got bitten by this before; we can't tell which variables in the config are non-nil, depending on which CLI commands instantiate it.

File: pkg/inject/inject.go (hunk 9 above)
Comment (ihcsim) [reply pkg/inject/inject.go]: Just a habit of grouping variables together.

File: pkg/charts/values_test.go (hunk 3 above)
Comment (ihcsim) [reply pkg/charts/values_test.go]: The injector doesn't use `values.yaml` during inject time. The `values.yaml` file is only relevant for `install`, when we work on https://github.com/linkerd/linkerd2/issues/3468. The reason for not defaulting `Trace` to nil here is because I'm hoping that this file is the only place where we will specify the default constant. 

> leading to a LINKERD2_PROXY_TRACE_COLLECTOR_SVC_NAME of default.serviceaccount.identity.$(_l5d_ns).$(_l5d_trustdomain) which is missing the namespace.

The tracing environment variables won't be added if the `CollectorSvcAddr` is empty.

Comment (ihcsim) [reply pkg/charts/values_test.go]: This change to the `values.yaml` is a leftover from pre- a4b17184add0b3f6b9579a459e32e9ad3aea1bdc. Since I assume we will tackle #3468 very soon, I figure it's a small enough change to be left alone.

Comment (adleong) [reply pkg/charts/values_test.go]: I can confirm that when the annotation is absent, the environment variables are also absent.
